src/components/RangeSlider/RangeSlider.js

```diff
@@ -0,0 +1,424 @@
+    const isAtTheSamePoint = bounds[closestBound + 1] === bounds[closestBound];
+
+    if (isAtTheSamePoint && bounds[recent] === bounds[closestBound]) {
+      boundNeedMoving = recent;
+    }
+
+    if (isAtTheSamePoint && value !== bounds[closestBound + 1]) {
+      boundNeedMoving =
+        value < bounds[closestBound + 1] ? closestBound : closestBound + 1;
+    }
+    return boundNeedMoving;
+  }
+
+  /**
+   * Returns an array of possible slider points, taking into account both
+   * and `step`. The result is cached.
+   */
+  getPoints() {
+    const { step, min, max } = this.props;
+    const cache = this.getPointsCache;
+    if (!cache || cache.step !== step) {
+      const pointsObject = {};
+      for (let point = min; point <= max; point += step) {
+        pointsObject[point] = point;
+      }
+      const points = Object.keys(pointsObject).map(parseFloat);
+      points.sort((a, b) => a - b);
+      this.getPointsCache = { step, points };
+    }
+    return this.getPointsCache.points;
+  }
+
+  getSliderStart() {
+    const slider = this.sliderRef;
+    const rect = slider.getBoundingClientRect();
+
+    return rect.left + global.window.pageXOffset;
+  }
+
+  getSliderLength() {
+    const slider = this.sliderRef;
+    if (!slider) {
+      return 0;
+    }
+
+    const coords = slider.getBoundingClientRect();
+    return coords.width;
+  }
+
+  handleGenerator = ({ index, ...restProps }) => {
+    if (restProps.value === null) {
+      return null;
+    }
+
+    return <Handle {...restProps} key={index} />;
+  };
+
+  moveTo(value) {
+    const { handle, bounds } = this.state;
+    const nextBounds = [...bounds];
+    nextBounds[handle] = value;
+
+    this.onChange({
+      handle,
+      bounds: nextBounds
+    });
+  }
+
+  calcValue(offset) {
+    const { min, max } = this.props;
+    const ratio = Math.abs(Math.max(offset, 0) / this.getSliderLength());
+    const value = ratio * (max - min) + min;
+    return value;
+  }
+
+  calcValueByPos(position) {
+    const { thershold } = this.props;
+    const { handle, bounds } = this.state;
+    const pixelOffset = position - this.getSliderStart();
+    const nextValue = utils.trimAlignValue(
+      this.calcValue(pixelOffset),
+      handle,
+      this.props,
+      bounds,
+      thershold
+    );
+    return nextValue;
+  }
+
+  addDocumentTouchEvents() {
+    global.window.addEventListener("touchmove", this.onTouchMove);
+    global.window.addEventListener("touchend", this.onEnd);
+  }
+
+  addDocumentMouseEvents() {
+    global.window.addEventListener("mousemove", this.onMouseMove);
+    global.window.addEventListener("mouseup", this.onEnd);
+  }
+
+  removeDocumentEvents() {
+    global.window.removeEventListener("touchmove", this.onTouchMove);
+    global.window.removeEventListener("touchend", this.onEnd);
+
+    global.window.removeEventListener("mousemove", this.onMouseMove);
+    global.window.removeEventListener("mouseup", this.onEnd);
+  }
```

Comment:
Do we need to add these event listeners to the dom? Can we get away with adding them to the `SliderContainer` instead?

Reply:
It is a must to add those listeners to the window because if they are added to the slider itself once we leave slider area (top/bottom) the events should be detached and the slider should stop moving even thought we did not release it. This is a bit confusing I believe we should not release the slider until we release it with mouse up/touch end instead of releasing it when we leave slider area, because it is possible to leave the area by accident on the top/bottom. Slider's height is not very big.